system-configuration/src/network_configuration.rs

```diff
@@ -0,0 +1,468 @@
+
+        SCNetworkServiceDns {
+            state_domain_name: state_domain_name,
+            state_server_addresses: state_server_addresses,
+            setup_domain_name: setup_domain_name,
+            setup_server_addresses: setup_server_addresses,
+        }
+    }
+
+    /// Setting DNS on this network service
+    pub fn set_dns(&self, dns: SCNetworkServiceDns) -> bool {
+        let store = SCDynamicStoreBuilder::new("ns_dns_set").build();
+
+        if dns.setup_server_addresses.is_some() {
+            let key = CFString::from_static_string("ServerAddresses");
+            let addrs: Vec<CFString> = dns.setup_server_addresses
+                .unwrap()
+                .iter()
+                .map(|s| CFString::new(&format!("{}", s)))
+                .collect();
+            let value = CFArray::from_CFTypes(&addrs);
+            let dictionary = CFDictionary::from_CFType_pairs(&[(key, value)]);
+
+            let path = CFString::new(&format!("Setup:/Network/Service/{}/DNS", self.id()));
+
+            if !store.set(path, dictionary) {
+                return false;
+            }
+        }
+
+        if dns.setup_domain_name.is_some() {
+            let key = CFString::from_static_string("DomainName");
+            let value = CFString::new(dns.setup_domain_name.unwrap().as_str());
+            let dictionary = CFDictionary::from_CFType_pairs(&[(key, value)]);
+
+            let path = CFString::new(&format!("Setup:/Network/Service/{}/DNS", self.id()));
+
+            if !store.set(path, dictionary) {
+                // FIXME: should rollback ?
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /// Returns the network interface associated with this network service.
+    pub fn interface(&self) -> Option<SCNetworkInterface> {
+        let interface_ptr = unsafe { SCNetworkServiceGetInterface(self.0) };
+        if interface_ptr.is_null() {
+            None
+        } else {
+            Some(unsafe { SCNetworkInterface::wrap_under_get_rule(interface_ptr) })
+        }
+    }
+}
+
+impl fmt::Display for SCNetworkService {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl fmt::Debug for SCNetworkService {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "SCNetworkService{{ id: {:?}, name: {:?}, enabled: {}, interface: {:?}, dns: {:?} }}",
+            self.id(),
+            self.name(),
+            self.enabled(),
+            self.interface(),
+            self.dns()
+        )
+    }
+}
+
+
+declare_TCFType!{
+    /// Network interface object.
+    SCNetworkInterface, SCNetworkInterfaceRef
+}
+
+impl_TCFType!(
+    SCNetworkInterface,
+    SCNetworkInterfaceRef,
+    SCNetworkInterfaceGetTypeID
+);
+
+
+impl SCNetworkInterface {
+    /// Returns primary network interface
+    pub fn global(store: &SCDynamicStore) -> Option<Self> {
```

Comment:
Could we not implement this method as `SCNetworkService::global(store).interface()`? That would avoid iterating all interfaces. Or can there be a case where the primary interface is not attached to a service, and thus that implementation would return something different from this one?

Reply:
that's make sense :)